modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientBinaryTupleUtils.java

```diff
@@ -49,72 +79,251 @@ public static void readAndSetColumnValue(
+            case TIMESTAMP:
                 tuple.set(columnName, reader.timestampValue(readerIndex));
                 break;
 
             default:
-                throw new IgniteException(PROTOCOL_ERR, "Unsupported type: " + clientDataType);
+                throw unsupportedTypeException(clientDataType);
         }
     }
+
+    /**
+     * Reads an object from binary tuple at the specified index.
+     *
+     * @param reader Binary tuple reader.
+     * @param index  Starting index in the binary tuple.
+     * @return Object.
+     */
+    public static Object readObject(BinaryTupleReader reader, int index) {
+        if (reader.hasNullValue(index)) {
+            return null;
+        }
+
+        int type = reader.intValue(index);
+        int valIdx = index + 2;
+
+        switch (type) {
+            case INT8:
+                return reader.byteValue(valIdx);
+
+            case INT16:
+                return reader.shortValue(valIdx);
+
+            case INT32:
+                return reader.intValue(valIdx);
+
+            case INT64:
+                return reader.longValue(valIdx);
+
+            case FLOAT:
+                return reader.floatValue(valIdx);
+
+            case DOUBLE:
+                return reader.doubleValue(valIdx);
+
+            case DECIMAL:
+                return reader.decimalValue(valIdx, reader.intValue(index + 1));
+
+            case UUID:
+                return reader.uuidValue(valIdx);
+
+            case STRING:
+                return reader.stringValue(valIdx);
+
+            case BYTES:
+                return reader.bytesValue(valIdx);
+
+            case BITMASK:
+                return reader.bitmaskValue(valIdx);
+
+            case DATE:
+                return reader.dateValue(valIdx);
+
+            case TIME:
+                return reader.timeValue(valIdx);
+
+            case DATETIME:
+                return reader.dateTimeValue(valIdx);
+
+            case TIMESTAMP:
+                return reader.timestampValue(valIdx);
+
+            case NUMBER:
+                return reader.numberValue(valIdx);
+
+            case BOOLEAN:
+                return reader.byteValue(valIdx) != 0;
+
+            case DURATION:
+                return reader.durationValue(valIdx);
+
+            case PERIOD:
+                return reader.periodValue(valIdx);
+
+            default:
+                throw unsupportedTypeException(type);
+        }
+    }
+
+    /**
+     * Writes an object with type info to the binary tuple.
+     *
+     * @param builder Builder.
+     * @param obj Object.
+     */
+    public static void appendObject(BinaryTupleBuilder builder, Object obj) {
+        if (obj == null) {
+            builder.appendNull();
+            builder.appendInt(0);
+            builder.appendNull();
+        } else if (obj instanceof Byte) {
+            builder.appendInt(ClientDataType.INT8);
+            builder.appendInt(0);
+            builder.appendByte((Byte) obj);
```

Comment:
Are those `type`, `scale` and `value`? Maybe we should add a function with 3 args just so it's more clear what is written.

Reply:
Good point, done.